### PR TITLE
Weddings

### PR DESCRIPTION
**Pull request by Cirras**
This PR implements weddings, based on the implementation from [mainclone-eoserv](https://github.com/eoserv/mainclone-eoserv/commit/e18e87f169140b6c0da80b9d866f672e8f6dced5).

Some changes have been made for accuracy:
- The minimum wedding level is enforced
- That [second prompt](https://user-images.githubusercontent.com/22178573/189476349-6a725fe9-001a-46d3-89c5-7c4f6761eb5b.png) at the start of the ceremony is implemented
- The fizz effect played during the ceremony uses the "map coordinate effect" instead of the "player effect" packet

https://user-images.githubusercontent.com/22178573/189476563-176fe641-39c7-425d-9699-b1ac05e896db.mp4

### CODE REVIEW

File: src/map.cpp
Comment (ethanmoffat) [inline src/map.cpp:363]: In looking into how things behave with `$remap`, I think there is a bug where arena gets unloaded and never reloaded. `LoadArena` is only ever called from the constructor but `Map::Reload` calls both `Unload` and `Load`.

I think we probably want to add a `LoadArena` call to the `Reload` call, move this to a `LoadWedding` method, and also call that in the `Reload` method. If the map mutates and the NPC is removed then there will be a wedding object for a nonexistent NPC.

Comment (ethanmoffat) [reply src/map.cpp:363]: Oh and also delete/clean up the wedding if the map mutates in the `Unload` call.

File: src/wedding.cpp
Comment (ethanmoffat) [inline src/wedding.cpp:64]: `Unregister` doesn't appear to delete the timer that gets passed to it. Does this get leaked?

**Review by ethanmoffat — COMMENTED** [review]
Implementation in `wedding.cpp` seems fine to me, looks like a bunch of edge cases beyond what I considered are already covered thanks to those packet response codes being well-documented.

I'll definitely spend some time testing this out, since it's kind of hard to verify anything just by reading the code.

File: config/misc.ini
Comment (Cirras) [reply config/misc.ini:170]: On the official game server, it was only Tuxedo for male characters and White Dress for female characters.

File: src/handlers/Marriage.cpp
Comment (Cirras) [reply src/handlers/Marriage.cpp:50]: Will leave this one to you, but your comment has made me realize that this is an unused variable.
I'll comment out the lhs, which is what's already done around the codebase for instances of session_id, dialog_id, spell_id, etc.

File: src/handlers/Marriage.cpp
Comment (Cirras) [reply src/handlers/Marriage.cpp:56]: Sure. I'll also use a switch statement.

File: src/handlers/Priest.cpp
Comment (Cirras) [reply src/handlers/Priest.cpp:28]: Right you are, I'll remove that.

File: src/handlers/Marriage.cpp
Comment (Cirras) [reply src/handlers/Marriage.cpp:60]: Same issue on several other packet builders in `Marriage.cpp`, `wedding.cpp`, and one `map.cpp`.
I'll correct them, thanks.

File: src/wedding.cpp
Comment (Cirras) [reply src/wedding.cpp:64]: Looks like it does, good catch.
Easy fix, simply deleting `tick_timer` should do the trick. A registered `TimeEvent` will unregister itself in the destructor.

**Comment by Cirras** [comment]
@ethanmoffat
Made requested changes.
If you're happy with them, I'll squash the commits into `Implement weddings` before merge.

File: config/misc.ini
Comment (ethanmoffat) [reply config/misc.ini:170]: I'd still lean toward offering a list of items as an option, and defaulting to what main did. nbd if it isn't part of this PR - I can pick that up once this merges.

File: src/map.cpp
Comment (Cirras) [reply src/map.cpp]: Oops, forgot to reply to this one.

Anyway, good observations here.
I've tidied up the map lifecycle issues around weddings and arena.

**Comment by ethanmoffat** [comment]
This looks good, thanks for getting to those changes so quickly. I'd like to test this out since I haven't played around with weddings/law too much. I'll run through a few cases tonight and see if I can break it and approve here when I inevitably can't.

Don't worry about squashing commits, I prefer a richer commit history, makes it easier to figure out why something was done.

**Comment by ethanmoffat** [comment]
@Cirras A couple `-Wreorder` warnings on Linux:

```
[ 32%] Building CXX object CMakeFiles/eoserv_lib.dir/tu/game_1.cpp.o
In file included from /home/ethan/git/etheos/tu/../src/map.cpp:21:0,
                 from /home/ethan/git/etheos/tu/game_1.cpp:9:
/home/ethan/git/etheos/tu/../src/wedding.hpp: In constructor ‘Wedding::Wedding(Map*, unsigned char)’:
/home/ethan/git/etheos/tu/../src/wedding.hpp:32:14: warning: ‘Wedding::tick_timer’ will be initialized after [-Wreorder]   TimeEvent* tick_timer;
              ^~~~~~~~~~
/home/ethan/git/etheos/tu/../src/wedding.hpp:23:8: warning:   ‘Map* Wedding::map’ [-Wreorder]
   Map* map;
        ^~~
In file included from /home/ethan/git/etheos/tu/game_1.cpp:14:0:
/home/ethan/git/etheos/tu/../src/wedding.cpp:106:1: warning:   when initialized here [-Wreorder]
 Wedding::Wedding(Map *map, unsigned char priest_idx)
 ^~~~~~~
[ 35%] Building CXX object CMakeFiles/eoserv_lib.dir/tu/game_2.cpp.o
```

File: src/handlers/Marriage.cpp
Comment (ethanmoffat) [inline src/handlers/Marriage.cpp:65]: Looks like `MarriageApproval` should be 1 and `Divorce` should be 0 in the enum

File: src/handlers/Priest.cpp
Comment (ethanmoffat) [inline src/handlers/Priest.cpp:112]: Upon further reflection these checks should probably be case-insensitive

File: src/handlers/Marriage.cpp
Comment (ethanmoffat) [inline src/handlers/Marriage.cpp:130]: We might want to add in sending this if someone gets divorced while their partner is offline, once they log back in. Not sure how complicated that would be though.

File: src/handlers/Marriage.cpp
Comment (Cirras) [reply src/handlers/Marriage.cpp:65]: Fixed.

File: src/handlers/Marriage.cpp
Comment (Cirras) [reply src/handlers/Marriage.cpp:130]: Probably unnecessary complexity, but I would absolutely agree with doing this if it's in the interest of emulating the original game server.
Could I get you to check the `GameServer.exe` behavior around this?

**Comment by Cirras** [comment]
> @Cirras A couple `-Wreorder` warnings on Linux:
> 
> ```
> [ 32%] Building CXX object CMakeFiles/eoserv_lib.dir/tu/game_1.cpp.o
> In file included from /home/ethan/git/etheos/tu/../src/map.cpp:21:0,
>                  from /home/ethan/git/etheos/tu/game_1.cpp:9:
> /home/ethan/git/etheos/tu/../src/wedding.hpp: In constructor ‘Wedding::Wedding(Map*, unsigned char)’:
> /home/ethan/git/etheos/tu/../src/wedding.hpp:32:14: warning: ‘Wedding::tick_timer’ will be initialized after [-Wreorder]   TimeEvent* tick_timer;
>               ^~~~~~~~~~
> /home/ethan/git/etheos/tu/../src/wedding.hpp:23:8: warning:   ‘Map* Wedding::map’ [-Wreorder]
>    Map* map;
>         ^~~
> In file included from /home/ethan/git/etheos/tu/game_1.cpp:14:0:
> /home/ethan/git/etheos/tu/../src/wedding.cpp:106:1: warning:   when initialized here [-Wreorder]
>  Wedding::Wedding(Map *map, unsigned char priest_idx)
>  ^~~~~~~
> [ 35%] Building CXX object CMakeFiles/eoserv_lib.dir/tu/game_2.cpp.o
> ```

Thanks, those warnings should be fixed now.

File: src/handlers/Priest.cpp
Comment (Cirras) [reply src/handlers/Priest.cpp:112]: I've lowercased the fiance name inputs instead, matching the actual casing of character names.
This is the approach in other places like `Account_Create`, `Guild_Player`, and `Guild_Rank`.

File: config/misc.ini
Comment (Cirras) [reply config/misc.ini:170]: This shouldn't be hard, I'll look at adding it.

File: src/handlers/Marriage.cpp
Comment (ethanmoffat) [reply src/handlers/Marriage.cpp:130]: Sure, I'll get back to you on that behavior

Comment (ethanmoffat) [reply src/handlers/Marriage.cpp:130]: `GameServer.exe` does not do this so I think we can skip it

**Comment by Cirras** [comment]
After watching your [recording](https://www.youtube.com/watch?v=3qH1WFVvyRg) of the weddings on `GameServer.exe`, I see a couple of minor opportunities for improvement:
- 1 second delay between clicking the "I do" prompt and actually saying "I do"
- Some minor dialogue differences

**Comment by Cirras** [comment]
@ethanmoffat
I think this is ready for another round of review.

**Review by ethanmoffat — APPROVED** [review]

**Comment by ethanmoffat** [comment]
This looks good. Going to do some quick testing and then merge it

**Pull request merged**
